catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -0,0 +1,502 @@
+                .getOrDefault(CONTENT_PATHS, new HashMap<>());
+
+        List<ContentItem> newContentItems = createStorageRequest.getContentItems()
+                .stream()
+                .map(ContentItem::getMetacard)
+                .distinct()
+                .flatMap(mc -> getVersionedContent(mc,
+                        createStorageRequest.getContentItems(),
+                        MetacardVersion.Action.CREATED_CONTENT,
+                        tmpContentPaths).stream())
+                .collect(Collectors.toList());
+
+        newContentItems.forEach(ci -> {
+            if (ci.getMetacard()
+                    .getResourceURI() == null) {
+                ci.getMetacard()
+                        .setAttribute(new AttributeImpl(Metacard.RESOURCE_URI, ci.getUri()));
+                try {
+                    ci.getMetacard()
+                            .setAttribute(new AttributeImpl(Metacard.RESOURCE_SIZE, ci.getSize()));
+                } catch (IOException e) {
+                    LOGGER.warn("Could not get size of content item", e);
+                }
+            }
+        });
+
+        CreateStorageRequestImpl versionStorageRequest = new CreateStorageRequestImpl(
+                newContentItems,
+                createStorageRequest.getId(),
+                createStorageRequest.getProperties());
+
+        storageProvider().create(versionStorageRequest);
+
+        // These are being staged for delayed execution, to be ran after the storage/catalog has
+        // successfully committed/updated
+        preStaged.add(() -> {
+            getSystemSubject().execute(() -> {
+                storageProvider().commit(versionStorageRequest);
+                return true;
+            });
+            return true;
+        });
+        preStaged.add(() -> {
+            getSystemSubject().execute(() -> {
+                catalogProvider().create(new CreateRequestImpl(newContentItems.stream()
+                        .map(ContentItem::getMetacard)
+                        .distinct()
+                        .collect(Collectors.toList())));
+                return true;
+            });
+            return true;
+        });
+
+        return Optional.of(transactionKey);
+    }
+
+    /**
+     * Versions updated {@link Metacard}s and {@link ContentItem}s.
+     *
+     * @param streamUpdateRequest   Needed to pass {@link MetacardVersion#SKIP_VERSIONING}
+     *                              flag into downstream update
+     * @param updateStorageResponse Versions this response's updated items
+     * @param tmpContentPaths       The temporary content paths needed to duplicate the
+     *                              {@link ContentItem}s
+     * @return The transaction key for using with {@link Historian#commit(String)} and
+     * {@link Historian#rollback(String)}
+     * @throws IOException
+     */
+    public Optional<String> versionUpdateStorage(UpdateStorageRequest streamUpdateRequest,
```

Comment:
This is fairly long method. Could you break it down into smaller steps to make it easier to understand?